### PR TITLE
Bug 1426307 - Store only derived keys instead of storing kB/kA.

### PR DESCRIPTION
**Pull request by eoger**
We derive the keys right after we unwrap kB.

### CODE REVIEW

File: Account/FxAState.swift [hunk 1]
```diff
@@ -80,36 +82,110 @@ func stateV1(fromJSON json: JSON) -> FxAState? {
...
                     let kB = json["kB"].string?.hexDecodedData,
                     let keyPairJSON = json["keyPair"].dictionaryObject,
                     let keyPair = RSAKeyPair(jsonRepresentation: keyPairJSON),
                     let keyPairExpiresAt = json["keyPairExpiresAt"].int64,
                     let certificate = json["certificate"].string,
                     let certificateExpiresAt = json["certificateExpiresAt"].int64 {
-                    return MarriedState(sessionToken: sessionToken, kA: kA, kB: kB,
-                        keyPair: keyPair, keyPairExpiresAt: UInt64(keyPairExpiresAt),
-                        certificate: certificate, certificateExpiresAt: UInt64(certificateExpiresAt))
+                        let kSync = FxAClient10.deriveKSync(kB)
```
Comment (ncalexan) [inline Account/FxAState.swift]: This indentation looks funky.

File: Account/FxAState.swift [hunk 2]
```diff
@@ -80,36 +82,110 @@ func stateV1(fromJSON json: JSON) -> FxAState? {
...
+
+            case .doghouse:
+                return DoghouseState()
+            }
+        }
+    }
+    return nil
+}
+
+// Same as V1 except we it contains kSync and kXCS instead of kB
```
Comment (ncalexan) [inline Account/FxAState.swift]: nit: fix comment ("we KNOW it")... and full sentence.  Better:

```// Identical to V1 except `(kA, kB)` have been replaced with `(kSync, kXCS)` throughout.)```

File: Account/FxAState.swift [hunk 3]
```diff
@@ -80,36 +82,110 @@ func stateV1(fromJSON json: JSON) -> FxAState? {
...
+func stateV2(fromJSON json: JSON) -> FxAState? {
```
Comment (ncalexan) [inline Account/FxAState.swift]: There's a lot of duplication here, but I guess that comes with the territory.  You could make this more clear by modifying the incoming `json` in place...

File: Account/FxAClient10.swift [hunk 4]
```diff
@@ -142,13 +142,17 @@ open class FxAClient10 {
...
-    open class func computeClientState(_ kB: Data) -> String? {
-        if kB.count != 32 {
-            return nil
-        }
+    open class func computeClientState(_ kB: Data) -> String {
         return kB.sha256.subdata(in: 0..<16).hexEncodedString
     }
 
+    open class func deriveKSync(_ kB: Data) -> Data {
+        let salt = Data()
```
Comment (garvankeeley) [inline Account/FxAClient10.swift]: Is this to generate random bytes? Is there a more explicit way to do this swift I wonder.

File: Account/FxAState.swift [hunk 5]
```diff
@@ -80,36 +82,110 @@ func stateV1(fromJSON json: JSON) -> FxAState? {
...
+    if let labelString = json["label"].string {
+        if let label = FxAStateLabel(rawValue: labelString) {
+            switch label {
+            case .engagedBeforeVerified:
+                if let
+                    sessionToken = json["sessionToken"].string?.hexDecodedData,
+                    let keyFetchToken = json["keyFetchToken"].string?.hexDecodedData,
+                    let unwrapkB = json["unwrapkB"].string?.hexDecodedData,
+                    let knownUnverifiedAt = json["knownUnverifiedAt"].int64,
+                    let lastNotifiedUserAt = json["lastNotifiedUserAt"].int64 {
```
Comment (garvankeeley) [inline Account/FxAState.swift]: Looks repeated in each `case`, consider pulling them out of the case body (if possible)

**Comment by eoger** [comment]
Oops I got used so much to rnewman reviewing my iOS patches that I put his nickname in r= instead of yours Nick, very sorry about that :(

**Pull request merged**